tensorflow/python/kernel_tests/array_ops/bitcast_op_test.py

```diff
@@ -33,7 +34,8 @@ def _testBitcast(self, x, datatype, shape):
       out = self.evaluate(tf_ans)
       buff_after = memoryview(out).tobytes()
       buff_before = memoryview(x).tobytes()
-      self.assertEqual(buff_before, buff_after)
+      if sys.byteorder == 'little' :
+        self.assertEqual(buff_before, buff_after)
```

Comment:
We shouldn't just skip the test.

I don't really agree with the previous PR, but it looks like we pretend the data is always little-endian prior to bitcasting, which means on BE machines we first do a copy to swap the byte order.  This test should therefore confirm this.  You can do this by byte-swapping one of the arrays before the buffer comparison, probably `out`.

Reply:
I believe buffer copy and swap on BE machines is only happening when we bitcast types of **different** sizes. So `out` ndarray might or might not be swapped. We can achieve the same logic in `bitcast_op_test_cpu` by simply comparing `x.itemsize == out.itemsize`.

Also, byte-swapping `out` would not work when we bitcast from bigger type to smaller. Let assume that we bit-casted a **int32** buffer(like BE buffer= 0x00 00 00 01) to _**four**_ **int8** buffer( which eventually becomes ndarray `out`= [1, 0, 0, 0]), if we do `out.byteswap()`, we would only be swapping those four int8 elements, which wont achieve anything neither it was our intention to swap single bytes.

I also tried to use [ndarray.astype()](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html) which is used to cast an array to a specific type, but turns out, it was not a low-level cast and it casted each element of the array to the specified type. For our case, it turned four int8 to four int32 elements.

I did manage to come up with a solution where we byte swap `out` ndarray when bitcasting from smaller to bigger type, and byte swap `x` ndarray(input buffer/array in our case) when bitcasting from bigger to smaller type. All tests are passing with these changes and these changes wont affect LE machines but i did check that these changes are passing on LE machines as well.

After changes **__testBitcast_** method would look something like this : 

```
  def _testBitcast(self, x, datatype, shape):
    with test_util.use_gpu():
      tf_ans = array_ops.bitcast(x, datatype)
      out = self.evaluate(tf_ans)
      buff_after = memoryview(out).tobytes()
      buff_before = memoryview(x).tobytes()
      if sys.byteorder == 'little' or x.itemsize == out.itemsize :
        self.assertEqual(buff_before, buff_after)
      else:
        if x.itemsize < out.itemsize :
          out = out.byteswap()
          buff_after = memoryview(out).tobytes()
        else:
          x = x.byteswap()
          buff_before = memoryview(x).tobytes()
        self.assertEqual(buff_before, buff_after)
      self.assertEqual(tf_ans.get_shape(), shape)
      self.assertEqual(tf_ans.dtype, datatype)
```

Could you please have a look at these changes and let us know how to proceed further.
Thank you very much!

Reply:
If the types are the same size, the buffer doesn't need to be copied since both will already be of the correct "endianness".  In all cases (whether a copy is made or not) the resulting output of the bit-cast operation should still be in BE form on BE machines. 

What we really want to test then is that the bytes in both input and output are identical when viewed as LE.  So you're right, you can't just swap `out`, you need to swap both.  You can (hopefully) just do:
```python
if sys.byteorder == 'little' :
  le_bytes_before = memoryview(x.byteswap()).tobytes()
  le_bytes_after = memoryview(out.byteswap()).tobytes()
else:
  le_bytes_before = memoryview(x).tobytes()
  le_bytes_after = memoryview(out).tobytes()

self.assertEqual(le_bytes_before, le_bytes_after)
```

Reply:
Thanks @cantonios , it does make perfect sense now!

I was assuming that same size bitcast needed to behave in some static manner for both BE and LE machines due to which copy and swap has been ignored for same sizes but copy and swap was never needed for this scenario as performing copy/swap on same size will give the same results as compared to not doing copy/swap operation on the copied buffer.
I was also a bit hesitant to change/corrupt input buffer in the test case for all cases. But after knowing bitcast behaviour on same sizes, input buffer will definitely be needed to be byte swapped as well.

I have made the necessary changes as you requested.
Could you please have a look.
Thank you very much!